Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli

```diff
@@ -38,71 +38,80 @@ float GetDetailClipmapScaleInv(uint clipmapLevel)
 }
 
-float2 GetCurrentMacroClipmapCenter(uint clipmapLevel)
+uint2 GetCurrentMacroClipmapCenter(uint clipmapLevel)
 {
     return TerrainSrg::m_clipmapData.m_macroClipmapCenters[clipmapLevel].zw;
 }
 
-float2 GetPreviousDetailClipmapCenter(uint clipmapLevel)
+uint2 GetPreviousDetailClipmapCenter(uint clipmapLevel)
 {
     return TerrainSrg::m_clipmapData.m_detailClipmapCenters[clipmapLevel].xy;
 }
 
-float2 GetCurrentDetailClipmapCenter(uint clipmapLevel)
+uint2 GetCurrentDetailClipmapCenter(uint clipmapLevel)
 {
     return TerrainSrg::m_clipmapData.m_detailClipmapCenters[clipmapLevel].zw;
 }
 
 // Get the world position at the pixel position in a clipmap.
 // Note: clipmap center is dynamic because we are using toroidal addressing. It's not always at the middle.
-float2 GetWorldPosition(float2 clipmapCenter, uint2 pixelPosition, uint clipmapLevel, float clipmapScaleInv, float maxRenderDistance)
+float2 GetWorldPosition(uint2 clipmapCenter, uint2 pixelPosition, uint clipmapLevel, float clipmapScaleInv, float maxRenderRadius)
 {
-    float2 normalizedPixelPosition = (pixelPosition.xy + 0.5) / TerrainSrg::m_clipmapData.m_clipmapSize;
-
-    float2 distance = normalizedPixelPosition - clipmapCenter;
-
-    // Toroidal addressing:
-    // If distance is out of the normalized range (-0.5, 0.5), meaning we need to start from the other side.
-    // The logic is equivalent to a modulation. Using step function for acceleration.
-    distance.x -= step(0.5, distance.x);
-    distance.x += step(distance.x, -0.5);
-    distance.y -= step(0.5, distance.y);
-    distance.y += step(distance.y, -0.5);
-
-    float2 viewRelativePosition = distance * (maxRenderDistance * 2.0f * clipmapScaleInv);
+    /*
+     ___________________________
+    |-2,-2 |      |      |      |   Taking a 4x4 clipmap as an example.
+    |      |      |      |      |   Due to discreteness, center is set at the bottom right of
+    |______|______|______|______|   the actual geometrical center,
+    |      |-1,-1 |      |      |   The distance to the clipmap center is ranged between
+    |      |      |      |      |   [-clipmapSize/2, clipmapSize/2 - 1].
+    |______|______|______|______|   This property fits the modulate operation, at shifted clipmapSize/2,
+    |      |      |center|      |   assuming the size of the clipmap is always even.
+    |      |      |0,0   |      |
+    |______|______|______|______|
+    |      |      |      |1,1   |
+    |      |      |      |      |
+    |______|______|______|______|
+
+    */
+    int2 distance = int2(pixelPosition) - int2(clipmapCenter);
+    int2 clipmapSize = int2(TerrainSrg::m_clipmapData.m_clipmapSizeUint, TerrainSrg::m_clipmapData.m_clipmapSizeUint);
+    int2 halfClipmapSize = clipmapSize / 2;
+    distance = (distance + halfClipmapSize + clipmapSize) % clipmapSize - halfClipmapSize;
+
+    float2 viewRelativePosition = distance * (maxRenderRadius * 2.0f * clipmapScaleInv / TerrainSrg::m_clipmapData.m_clipmapSizeFloat);
```

Comment:
Because radius is half of the clipmap size. The number is easier to understand in the configuration. I'll put this into a function and add comments to explain.